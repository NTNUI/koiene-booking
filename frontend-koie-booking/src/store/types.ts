// rootState = generic type, defines the root state type
import AdminBooking, { AdminBookingDictionary } from '@/types/admin/AdminBooking';
import { AdminReportKoieDictionary } from '@/types/admin/AdminReports';

export interface RootState {
  version: string;
  koie: KoieState;
  booking: BookingState;
  auth: AuthState;
  avalanche: AvalancheState;
  adminBookings: AdminBookingsState;
}

export interface KoieState {
  currentLocale: string;
  isLoading: boolean;
  koieData: Koie;
  error: boolean;
  allKoier: [];
}

export interface Koie {
  location: {
    latitude: string;
    longitude: string;
    area: string;
    difficultyInfo_eng: string;
    difficultyInfo_nor: string;
    altitude: number;
    terrain_eng: string;
    terrain_nor: string;
    map_pdf: string;
    kartblad: string;
  };
  description: {
    yr_link_eng: string;
    yr_link_nor: string;
    description_eng: string;
    description_nor: string;
    directions_eng: string;
    directions_nor: string;
    parking_eng: string;
    parking_nor: string;
  };
  beds_available_in_booking_window: Array<BedsAvailable>;
  name: string;
  number_of_beds: number;
  booking_window: number;
  price_member: number;
  price_not_member: number;
}

export interface BedsAvailable {
  date: string;
  bedsAvailable: number;
}

export interface BookingState {
  isLoading: boolean;
  dateFrom: string;
  dateTo: string;
  beds: number;
  step: number;
  guests: Array<Guest>;
  numberOfMembers: number;
  numberOfNonMembers: number;
  edited: boolean;
  validForm: boolean;
  unavailableDates: [];
  availableBeds: number;
  noDatesAvailable: boolean;
  error: boolean;
  bookingData: BookingInfo;
  paymentConfirmed: boolean;
}

export interface ReportState {
  validForm: boolean;
  edited: boolean;
  isLoading: boolean;
  error: boolean;
  step: number;
  reportData: ReportData;
  bookingInfo: BookingInfo;
}

export interface ReportData {
  id: number;
  booking_uuid: string;
  firewood: number;
  chopped_up_wood: number;
  smoke_detector_is_working: boolean;
  gas_is_full: boolean;
  gas_burner_primus: number;
  axe: number;
  hammer: number;
  saw: number;
  saw_blade: number;
  saw_bench: number;
  spade: number;
  kerosene_lamp: number;
  detergent: number;
  dishware: number;
  cookware: number;
  cabin_book: number;
  candle_holders: number;
  fire_blanket: number;
  fire_extinguisher: number;
  other_faults: string;
  boat_status: number;
  canoe_status: number;
  life_jackets_status: number;
  feedback: string;
}

export interface AvalancheState {
  warningLevels: Array<Warning>;
  disableBooking: boolean;
  warningData: Array<WarningResult>;
  koierToCheck: Array<string>;
}

export interface Warning {
  day: string;
  level: number;
}

export interface WarningResult {
  dangerLevel: number;
  validTo: string;
}

export interface CreateBookingInfo {
  koie: string;
  from_date: string;
  to_date: string;
  guests: Array<Guest>;
  guests_member: number;
  guests_not_member: number;
}

export interface BookingInfo {
  id: number;
  uuid: string;
  user: number;
  koie: string;
  price: number;
  booking_transaction_id: string;
  from_date: string;
  to_date: string;
  guests_member: number;
  guests_not_member: number;
}

export interface AuthState {
  isAdmin: boolean;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
}

export interface Guest {
  name: string;
  number: string;
  email: string;
  isMember: boolean;
}

export interface AdminBookingsState {
  cabinsWithBookings: AdminBookingDictionary;
  startDate: string;
}

export interface AdminReportsState {
  chosenCabin: string;
  expanded: boolean;
  cabins: AdminReportKoieDictionary;
}
