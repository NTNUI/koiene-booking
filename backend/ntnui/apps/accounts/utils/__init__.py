from accounts.utils.mail_utils import send_gsuite_confirmation_mail
