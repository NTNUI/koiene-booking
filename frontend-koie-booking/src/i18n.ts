import Vue from 'vue';
import VueI18n, { LocaleMessages } from 'vue-i18n';
import enJson from './locales/en.json';

Vue.use(VueI18n);

function loadLocaleMessages(): LocaleMessages {
  // Testing currently only supports manually importing of json-files. English is the only language that gets loaded now
  let locales: any;
  const messages: LocaleMessages = {};
  if (process.env.NODE_ENV === 'test') {
    messages['en'] = enJson;
  } else {
    locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
    locales.keys().forEach((key: any) => {
      const matched = key.match(/([A-Za-z0-9-_]+)\./i);
      if (matched && matched.length > 1) {
        const locale = matched[1];
        messages[locale] = locales(key);
      }
    });
  }
  return messages;
}

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages()
});
