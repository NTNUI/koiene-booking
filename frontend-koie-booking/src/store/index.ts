import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import axios from 'axios';
import VueAxios from 'vue-axios';
import { RootState } from './types';
import { koie } from './modules/users/koie/index';
import { booking } from './modules/users/booking/index';
import { avalanche } from './modules/users/avalanche/index';
import { report } from './modules/users/report/index';
import { auth } from './modules/users/auth/index';
import { adminBookings } from '@/store/modules/admin/bookings';

Vue.use(Vuex, VueAxios, axios);

const store: StoreOptions<RootState> = {
  modules: {
    adminBookings,
    koie,
    booking,
    avalanche,
    report,
    auth
  }
};

export default new Vuex.Store<RootState>(store);

export { store as storeConfig };
