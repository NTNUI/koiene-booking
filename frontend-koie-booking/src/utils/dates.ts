import dayjs from 'dayjs';
require('dayjs/locale/nb');
dayjs.locale('nb');

export function getDateString(baseDate?: string, addDays?: number): string {
  let resultString: string;
  if (baseDate && addDays) {
    resultString = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else if (baseDate) {
    resultString = new Date(baseDate).toISOString().substr(0, 10);
  } else if (addDays) {
    resultString = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else {
    resultString = new Date().toISOString().substr(0, 10);
  }
  return resultString;
}

export function getDate(baseDate?: string, addDays?: number): Date {
  let resultDate: Date;
  if (baseDate && addDays) {
    resultDate = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000);
  } else if (baseDate) {
    resultDate = new Date(baseDate);
  } else if (addDays) {
    resultDate = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000);
  } else {
    resultDate = new Date();
  }
  return resultDate;
}

export function addToDate(startDate: string, howMany: number, timeUnit: dayjs.OpUnitType): string {
  const date = dayjs(startDate + 'T00:00:00.000Z');
  return date.add(howMany, timeUnit).format('YYYY-MM-DD');
}

export function formatDate(dateISO: string, formatString: string): string {
  const date = dayjs(dateISO + 'T00:00:00.000Z');
  return date.format(formatString);
}
