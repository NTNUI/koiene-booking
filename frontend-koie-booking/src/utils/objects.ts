export const clone = (object: any) => Object.assign({}, object);

export const renameKey = (object: any, key: string, newKey: string): any => {
  const value = object[key];
  delete object[key];
  object[newKey] = value;
  return object;
};

export const sortObjectByKey = (unordered: any) => {
  let ordered: any;
  ordered = {};
  Object.keys(unordered)
    .sort()
    .forEach(function(key) {
      ordered[key] = unordered[key];
    });
  return ordered;
};
