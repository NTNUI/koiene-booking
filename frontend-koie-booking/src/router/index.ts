import Vue from 'vue';
import VueRouter from 'vue-router';
import store from '@/store';

Vue.use(VueRouter);

const routes = [
  {
    path: '/',
    name: 'AllKoier',
    component: () => import(/* webpackChunkName: "about" */ '../views/AllKoier.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/koie/:id',
    name: 'koie',
    component: () => import('../views/Koie.vue'),
    props: true,
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/booking/:id',
    name: 'booking',
    component: () => import('../views/Booking.vue'),
    meta: { requiresAuth: true, requiresAdmin: false }
  },
  {
    path: '/login',
    name: 'Login',
    component: () => import('../views/Login.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/admin',
    name: 'Admin',
    component: () => import('../views/Admin.vue'),
    meta: { requiresAuth: true, requiresAdmin: true }
  }
];

const router = new VueRouter({
  routes
});

export function beforeEach(to: any, from: any, next: any) {
  const isLoggedIn = store.getters['auth/isLoggedIn'];
  const isAdmin = store.getters['auth/isAdmin'];
  const requiresAuth = to.matched.some((record: any) => record.meta.requiresAuth);
  const requiresAdmin = to.matched.some((record: any) => record.meta.requiresAdmin);

  if (requiresAdmin && !isAdmin) next({ name: 'Login', query: { redirect: to.path } });
  else if (requiresAuth && !isLoggedIn) next({ name: 'Login', query: { redirect: to.path } });
  else if (to.name === 'Login' && isLoggedIn) next('/');
  else next();
}

router.beforeEach((to, from, next) => beforeEach(to, from, next));

export default router;
export { routes };
