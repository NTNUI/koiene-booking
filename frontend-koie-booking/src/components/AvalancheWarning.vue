<template>
  <v-layout v-model="dateFrom" width="100%" :class="$style.avalanche">
    <div :class="$style.warning">
      <v-avatar :class="$style.avatar" :tile="true" :color="colors[levelFirstDay]">
        <span class="white--text headline">{{ levelFirstDay }}</span>
      </v-avatar>
      <div :v-model="heading1" :class="$style.day">
        <h4>{{ heading1 }}</h4>
        <p class="mb-0">{{ warningTypes[levelFirstDay] }}</p>
      </div>
    </div>
    <div :class="$style.warning">
      <v-avatar :class="$style.avatar" :tile="true" :color="colors[levelSecondDay]">
        <span class="white--text headline">{{ levelSecondDay }}</span>
      </v-avatar>
      <div :class="$style.day">
        <h4>{{ heading2 }}</h4>
        <p class="mb-0">{{ warningTypes[levelSecondDay] }}</p>
      </div>
    </div>
    <div :class="$style.warning">
      <v-avatar :class="$style.avatar" :tile="true" :color="colors[levelThirdDay]">
        <span class="white--text headline">{{ levelThirdDay }}</span>
      </v-avatar>
      <div :class="$style.day">
        <h4>{{ heading3 }}</h4>
        <p class="mb-0">{{ warningTypes[levelThirdDay] }}</p>
      </div>
    </div>
  </v-layout>
</template>

<script lang="ts">
import Vue from 'vue';
import { AvalancheData } from '../types/avalanche';
import { getDateString, getDate } from '@/utils/dates';

export default Vue.extend({
  name: 'AvalancheWarning',
  props: {
    latitude: {
      type: String,
      default: ''
    },
    longitude: {
      type: String,
      default: ''
    }
  },

  data(): AvalancheData {
    return {
      result: '',
      levelFirstDay: '',
      levelSecondDay: '',
      levelThirdDay: '',
      firstDay: '',
      secondDay: '',
      thirdDay: '',
      fromDate: getDateString(),
      nextDate: getDateString(),
      lastDate: getDateString(),
      warningTypes: [
        this.$i18n.t('avalanche.warningTypes.low'),
        this.$i18n.t('avalanche.warningTypes.moderate'),
        this.$i18n.t('avalanche.warningTypes.considerable'),
        this.$i18n.t('avalanche.warningTypes.high'),
        this.$i18n.t('avalanche.warningTypes.very_high'),
        this.$i18n.t('avalanche.warningTypes.avalanche_expected')
      ],
      colors: [
        this.$scssVars.globalColorWarningLow,
        this.$scssVars.globalColorWarningModerate,
        this.$scssVars.globalColorWarningConsiderable,
        this.$scssVars.globalColorWarningHigh,
        this.$scssVars.globalColorWarningVeryHigh,
        this.$scssVars.globalColorBlack
      ]
    };
  },
  computed: {
    dateFrom(): string {
      this.findAllDates();
      return this.$store.state.booking.dateFrom;
    },
    dateTo(): string {
      //if you have only chosen one night, the warning will still show for two next days
      if (this.findAllDates().length < 3) {
        return this.lastShownWarningDate;
      } else {
        return this.$store.state.booking.dateTo;
      }
    },

    heading1(): string {
      return this.firstDay + ' - ' + this.fromDate.slice(0, 2) + '.' + this.fromDate.slice(3, 5);
    },
    heading2(): string {
      return this.secondDay + ' - ' + this.nextDate.slice(0, 2) + '.' + this.nextDate.slice(3, 5);
    },
    heading3(): string {
      return this.thirdDay + ' - ' + this.lastDate.slice(0, 2) + '.' + this.lastDate.slice(3, 5);
    },
    lastShownWarningDate(): string {
      const fromDateTime = this.$store.state.booking.dateFrom
        ? new Date(this.$store.state.booking.dateFrom).getTime()
        : new Date().getTime();
      return new Date(fromDateTime + 2 * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
    }
  },
  watch: {
    dateFrom: function() {
      this.setAvalancheWarning();
    },
    dateTo: function() {
      this.setAvalancheWarning();
    }
  },

  mounted() {
    this.setAvalancheWarning();
  },
  methods: {
    findAllDates(): Array<string> {
      const bookingDays = [];
      const fromDate = getDate(this.$store.state.booking.dateFrom);
      const toDate = getDate(this.$store.state.booking.dateTo);
      let bookingDaysInTime = toDate.getTime() - fromDate.getTime();
      let numberOfBookingDays = bookingDaysInTime / (1000 * 3600 * 24);
      for (let i = 0; i <= numberOfBookingDays; i++) {
        bookingDays.push(new Date(fromDate.getTime() + i * 24 * 60 * 60 * 1000).toISOString().substr(0, 10));
      }
      return bookingDays;
    },
    async setAvalancheWarning() {
      const language = 2;
      const values = {
        latitude: this.$store.state.koie.koieData.location.latitude,
        longitude: this.$store.state.koie.koieData.location.longitude,
        language: language,
        dateFrom: this.dateFrom,
        dateTo: this.dateTo
      };
      //returns avalance warning from varsom.no
      try {
        await this.$store.dispatch('avalanche/FETCH_AVALANCHE_LEVELS', values);
        const warningData = this.$store.state.avalanche.warningData;

        this.levelFirstDay = warningData[0].warningLevel;
        this.levelSecondDay = warningData[1].warningLevel;
        this.levelThirdDay = warningData[2].warningLevel;
        const dates = this.findAllDates();
        this.setAvalancheStore(warningData, dates);
        this.firstDay = getDateString(warningData[0].validTo);
        this.setDateDay(warningData);
        this.getFormattedDate(warningData);
      } catch (error) {
        throw error;
      }
    },
    setAvalancheStore(res: any, dates: Array<string>) {
      const warningLevels = [];
      for (let date in dates) {
        let warningObject = { day: '', level: 0 };
        warningObject.day = dates[date];
        warningObject.level = res[date].warningLevel;
        warningLevels.push(warningObject);
      }
      this.$store.dispatch('avalanche/SET_WARNING_LEVELS', warningLevels);
      for (let e in warningLevels) {
        if (warningLevels[e].level > 2) {
          this.$store.dispatch('avalanche/DISABLE_BOOKING', true);
          return;
        } else {
          this.$store.dispatch('avalanche/DISABLE_BOOKING', false);
        }
      }
    },
    //this method will set the weekday of the date
    setDateDay(res: any) {
      const weekdays = [
        this.$i18n.t('avalanche.weekdays.sunday'),
        this.$i18n.t('avalanche.weekdays.monday'),
        this.$i18n.t('avalanche.weekdays.tuesday'),
        this.$i18n.t('avalanche.weekdays.wednesday'),
        this.$i18n.t('avalanche.weekdays.thursday'),
        this.$i18n.t('avalanche.weekdays.friday'),
        this.$i18n.t('avalanche.weekdays.saturday')
      ];
      this.firstDay = getDateString(res[0].validTo);
      const todayDayInWeek = new Date(this.firstDay).getDay();
      this.firstDay = weekdays[todayDayInWeek].toString();
      //checks if today is friday or later, in that case list needs to start over
      if (todayDayInWeek === 5) {
        this.secondDay = weekdays[todayDayInWeek + 1].toString();
        this.thirdDay = weekdays[0].toString();
      } else if (todayDayInWeek === 6) {
        this.secondDay = weekdays[0].toString();
        this.thirdDay = weekdays[1].toString();
      } else {
        this.secondDay = weekdays[todayDayInWeek + 1].toString();
        this.thirdDay = weekdays[todayDayInWeek + 2].toString();
      }
    },
    //this method will format dates into "13.02" for display in avalanche warning
    getFormattedDate(res: any) {
      this.fromDate = getDateString(res[0].validTo);
      this.nextDate = getDateString(res[1].validTo);
      this.lastDate = getDateString(res[2].validTo);
      this.fromDate = this.fromDate
        .split('-')
        .reverse()
        .toString();
      this.nextDate = this.nextDate
        .split('-')
        .reverse()
        .toString();
      this.lastDate = this.lastDate
        .split('-')
        .reverse()
        .toString();
    }
  }
});
</script>

<style lang="scss" module>
.avalanche {
  display: flex;
  flex-direction: column;
  width: 100%;
}
.avatar {
  border-radius: 4px;
  margin: 5px;
}
.warning {
  border-top: solid;
  border-width: 1px;
  border-color: lightgrey;
  width: 100%;
  display: flex;
  flex-direction: row;
  padding: 5px;
}

.day {
  display: flex;
  flex-direction: column;
  width: 100%;
  margin-left: 10px;
}
</style>
