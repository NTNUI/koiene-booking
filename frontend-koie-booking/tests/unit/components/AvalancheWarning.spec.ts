import Vue from 'vue';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';
import Vuex from 'vuex';
import { storeConfig } from '@/store';
import { cloneDeep } from 'lodash';
import i18n from '@/i18n';
import scssVars from '@/styles/variables.scss';
import flushPromises from 'flush-promises';
import mockAxios from 'jest-mock-axios';
import { routes } from '@/router';
import { getDateString } from '@/utils/dates';

Vue.use(Vuetify);

// Utilities
import { mount, createLocalVue } from '@vue/test-utils';

// Components or views
import AvalancheWarning from '../../../src/components/AvalancheWarning.vue';

describe('Component AvalancheWarning.vue', () => {
  const router = new VueRouter({ routes, mode: 'abstract' });
  let wrapper: any;
  let localVue: any;
  let vuetify: any;
  let store: any;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    localVue.use(VueRouter);
    localVue.use(Vuetify);
    localVue.use(Vuex);
    // Hard resets the store between tests
    store = new Vuex.Store(cloneDeep(storeConfig));
    localVue.prototype.$scssVars = scssVars;
  });

  afterEach(() => {
    mockAxios.reset();
  });

  it('Matches snapshot', async () => {
    wrapper = mount(AvalancheWarning, {
      localVue,
      router,
      vuetify,
      i18n,
      store,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('Disabling booking if dangerLevel >2', async () => {
    const dates = [new Date().toISOString(), getDateString(undefined, 1), getDateString(undefined, 2)];
    store.dispatch('booking/SET_DATE_FROM', dates[0]);
    store.dispatch('booking/SET_DATE_TO', dates[2]);

    wrapper = mount(AvalancheWarning, {
      localVue,
      router,
      vuetify,
      i18n,
      store,
    });

    const spy = jest.spyOn(wrapper.vm.$store, 'dispatch');

    const responseObj = {
      data: [
        { DangerLevel: '3', ValidTo: dates[0] },
        { DangerLevel: '3', ValidTo: dates[1] },
        { DangerLevel: '3', ValidTo: dates[2] },
      ],
    };
    mockAxios.mockResponse(responseObj);
    await flushPromises();

    expect(spy).toBeCalledWith('avalanche/DISABLE_BOOKING', true);
    expect(spy).not.toBeCalledWith('avalanche/DISABLE_BOOKING', false);
  });
});
