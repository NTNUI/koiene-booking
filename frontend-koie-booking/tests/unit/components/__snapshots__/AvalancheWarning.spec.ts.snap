// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component AvalancheWarning.vue Matches snapshot 1`] = `
<div class="layout width value avalanche">
  <div class="warning">
    <div class="v-avatar rounded-0 avatar" style="height: 48px; min-width: 48px; width: 48px;"><span class="white--text headline"></span></div>
    <div v-model=" - 20.0-" class="day">
      <h4> - 20.0-</h4>
      <p class="mb-0"></p>
    </div>
  </div>
  <div class="warning">
    <div class="v-avatar rounded-0 avatar" style="height: 48px; min-width: 48px; width: 48px;"><span class="white--text headline"></span></div>
    <div class="day">
      <h4> - 20.0-</h4>
      <p class="mb-0"></p>
    </div>
  </div>
  <div class="warning">
    <div class="v-avatar rounded-0 avatar" style="height: 48px; min-width: 48px; width: 48px;"><span class="white--text headline"></span></div>
    <div class="day">
      <h4> - 20.0-</h4>
      <p class="mb-0"></p>
    </div>
  </div>
</div>
`;
